Kubelet enabling to support pod-overhead

egernst opened the pull request:
Pod and burstable QoS cgroups should take overhead of running a sandbox
into account if the PodOverhead feature is enabled. Update helper
functions which are utilized by Kubelet for sizing the pod and burstable QoS
cgroups as well as for eviction and preemption handling.

While enabling PodOverhead support, refactored a couple of functions to remove duplicated logic and improve test coverage (ie, have eviction handling use the resources package).

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Pod and burstable QoS cgroups should take overhead of running a sandbox
into account if the PodOverhead feature is enabled. Similarly, pod overhead should be considered within the eviction handling in Kubernetes.  

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:


```release-note
Introduce support for applying pod overhead to pod cgroups, if the PodOverhead feature is enabled.
```

Commit:
pod-overhead: utilize pod overhead for cgroup sizing, eviction handling

Pod and burstable QoS cgroups should take overhead of running a sandbox
into account if the PodOverhead feature is enabled. These helper
functions are utilized by Kubelet for sizing the pod and burstable QoS
cgroups.

Pod overhead is added to resource requests, regardless of the initial
request values. A particular resource pod overhead is only added to a
resource limit if a non-zero limit already existed.

This commit updates eviction handling to also take Pod Overhead into
account (if the feature is enabled).

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

egernst commented:
/cc @tallclair @dashpole 

The changes are pretty basic, but I was having issues getting a basic unit test working.  AFAICT, when you add two resourceLists, the string associated with each  quantity is dropped.

I *think* this is why the Deep.Equal is failing.  Is this the expected behavior?  Or, any suggestions?

egernst commented on pkg/api/v1/resource/helpers_test.go:
```diff
@@ -216,6 +220,97 @@ func TestExtractResourceValue(t *testing.T) {
+				v1.ResourceName(v1.ResourceMemory): resource.MustParse("5"),
+			},
+			expectedLimits: v1.ResourceList{},
+		}, /* Failing Test - TODO/help
```
This is the test case where I'm seeing issues -- this is the only point in the tests where we are actually adding Resource Lists.  From looking at the Add function, it seems the quantities string is cleared out, and that's the delta I see for the failure.  

Is there a best-practice for testing equality of resource lists?

dashpole commented:
Can we use quantity's `Equal` method to test equality of quantities instead of DeepEqual?

egernst commented:
@dashpole - yes, but that'll require looping through each resource in the list, which isn't ideal.  Will give it a shot...

egernst commented:
Updated the test per @dashpole's recommendation, and improved limits handling

egernst commented:
/cc @mcastelino @jcvenegas

Commit:
kube-eviction: use common resource summation functions

Utilize resource helpers' GetResourceRequestQuantity instead of
duplicating the logic here.

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

Commit:
resource: modify resource helpers for better reuse

Update GetResoureqRequest function to utilize a new helper,
GetResourceRequestQuantity. This logic was duplicated in a couple of
areas in the K/K codebase, so consolidating for better test coverage.

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

egernst commented:
/assign @dashpole @erictune

egernst commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -27,6 +27,7 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/klog"
+	v1resource "k8s.io/kubernetes/pkg/api/v1/resource"
```
@tallclair - my apologies - it wasn't 100% clear to me if anything should be rectified in kubernetes/pkg/api/resource as well (the helpers.go looks the same as what's in v1, but out of date).  Some of the versioning API is still a bit magical to me; can you clarify?

tallclair replied:
https://github.com/kubernetes/kubernetes/pull/81104 :smile:

tallclair replied:
This was very triggering to me. Look what you did: https://github.com/kubernetes/kubernetes/pull/81189
https://github.com/kubernetes/kubernetes/pull/81206

egernst commented:
I could use help adding this to the 1.16 milestone

egernst commented:
Current push is wrong in that it modifies the QoS by adding an overhead to requests, even if no requests were provided before. Will fix up on Monday...

tallclair commented on pkg/api/v1/resource/helpers_test.go:
```diff
@@ -241,6 +273,80 @@ func TestExtractResourceValue(t *testing.T) {
+	}{
+		{
+			cName:            "just-limit-no-overhead",
+			pod:              getOverheadPod("foo", "", "9", "", "", "", ""),
```
same here: re  long args list

egernst commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +68,75 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
+
+	// if PodOverhead feature is supported, add overhead for running a pod
+	// to the non-zero sum of container requests/limits:
+	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
```
There were concerns, which surprised me.  We ended up keeping it as is, since it is first checked against the Overhead being a part of the pod.Spec, anyway.  It'd be abnormal to hit this check and have Overhead set and the feature gate not set.

egernst commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +68,75 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
 	}
+
+	// if PodOverhead feature is supported, add overhead for running a pod
+	// to the non-zero sum of container requests/limits:
```
I had gone back and forth on this.  I think it makes sense not to add it, since it would effectively be effectively changing the QoS, right?

While the original spec would have this field as best-effort, we'd be providing requests, slightly changing the eviction behavior between a pod utilizing pod overhead and not.

I've gone back and forth, so am happy to hear your input on this.

egernst replied:
addressed

dashpole commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +68,75 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
+// GetResourceRequestQuantity finds and returns the request quantity for a specific resource.
+func GetResourceRequestQuantity(pod *v1.Pod, resourceName v1.ResourceName) resource.Quantity {
+
+	requestQuantity := resource.Quantity{Format: resource.BinarySI}
```
Can we switch the Format based on resource, and default to `resource.DecimalSI` if it is a resource that it isn't in types/v1?  It doesn't matter for eviction, but will be useful if anyone else uses the function.

egernst replied:
ACK.  

FYI: opened this issue, as I see Format being set inconsistently (in tests only): https://github.com/kubernetes/kubernetes/issues/81379

egernst replied:
addressed

dashpole reviewed (commented):
Really helps to clean things up.

egernst commented on pkg/api/v1/resource/helpers_test.go:
```diff
@@ -241,6 +273,80 @@ func TestExtractResourceValue(t *testing.T) {
+	}
+}
+
+func podOverheadResourcesEqual(res1, res2 v1.ResourceList) bool {
```
Thanks, much better than making my own!

Commit:
resource: cleanup helpers_test.go

No test content changes - just improvements for readability

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

egernst reviewed (commented):
Updated based on feedback - PTAL @tallclair @dashpole

tallclair commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,86 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
+	// if PodOverhead feature is supported, add overhead for running a pod
+	// to the sum of reqeuests and to non-zero limits:
+	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
+
```
nit: extra newline

In general, don't add a newline after opening a block.

tallclair commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,86 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
-		return 1
+// GetResourceRequestQuantity finds and returns the request quantity for a specific resource.
+func GetResourceRequestQuantity(pod *v1.Pod, resourceName v1.ResourceName) resource.Quantity {
+
```
nit: extra newline (also after opening a block)

tallclair commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,86 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
+		requestQuantity = resource.Quantity{Format: resource.BinarySI}
+	default:
+		requestQuantity = resource.Quantity{Format: resource.DecimalSI}
+
```
nit: extra newline

You can add a newline _after_ closing a block, but not before the closing block.

tallclair commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,86 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
+	switch resourceName {
+	case v1.ResourceCPU:
+		requestQuantity = resource.Quantity{Format: resource.DecimalSI}
+	case v1.ResourceMemory:
```
nit: prefer
```suggestion
	case v1.ResourceMemory, v1.ResourceStorage, v1.ResourceEphemeralStorage:
```

egernst replied:
Agreed.  Thanks!

tallclair commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,86 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
-				totalResources = rQuantity.Value()
+		if rQuantity, ok := container.Resources.Requests[resourceName]; ok {
+			if requestQuantity.Cmp(rQuantity) < 0 {
+				requestQuantity = rQuantity
```
reusing rQuantity here could result in adding the podOverhead to the cached resource request. Better to use `.Copy()`, `.DeepCopy()` (what's the difference?) or `.Set(...)`

egernst replied:
Done

tallclair commented on pkg/api/v1/resource/helpers_test.go:
```diff
@@ -64,27 +68,54 @@ func TestDefaultResourceHelpers(t *testing.T) {
 }
 
 func TestGetResourceRequest(t *testing.T) {
+
```
nit: extra newline

tallclair commented on pkg/api/v1/resource/helpers_test.go:
```diff
@@ -241,37 +273,125 @@ func TestExtractResourceValue(t *testing.T) {
+	}
+}
+
+type podResources struct {
```
thanks, so much more readable this way.

tallclair reviewed (commented):
Mostly nits, sorry to complain about the newlines.

egernst commented:
Thanks @tallclair - will keep that in mind for next PR.  Addressed feedback, including utilizing .Copy() instead of assigning potential cached value.

tallclair commented:
Thanks!

/lgtm
/approve

Need to fix up ssome of those CI failures. rerun `hack/update-bazel.sh`, not sure if there are other issues.

tallclair commented:
/lgtm

tallclair commented:
/assign @thockin 
For `/pkg/api` approval.

thockin reviewed (commented):
Thanks!

/approve

egernst commented:
/test pull-kubernetes-e2e-gce

tallclair commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,79 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
-				totalResources = rQuantity.Value()
+		if rQuantity, ok := container.Resources.Requests[resourceName]; ok {
+			if requestQuantity.Cmp(rQuantity) < 0 {
+				requestQuantity = *rQuantity.Copy()
```
nit: use DeepCopy (https://github.com/kubernetes/kubernetes/pull/81627)
```suggestion
				requestQuantity = rQuantity.DeepCopy()
```

egernst replied:
done

tallclair commented:
Hmm, the test failures are somewhat suspicious since it's the density & loader tests failing, which would be affected by PodOverhead. Except that overhead on those should be zero, and I think the feature gate is disabled, so probably a flake. Worth taking a look at though.

egernst commented:
> Hmm, the test failures are somewhat suspicious since it's the density & loader tests failing, which would be affected by PodOverhead. Except that overhead on those should be zero, and I think the feature gate is disabled, so probably a flake. Worth taking a look at though.

Thanks. It is at the very least a reminder of e2e tests which should be added for PodOverhead.

tallclair commented:
/lgtm

tedyu commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -65,37 +69,79 @@ func PodRequestsAndLimits(pod *v1.Pod) (reqs, limits v1.ResourceList) {
-				totalResources += rQuantity.Value()
-			}
+		if rQuantity, ok := container.Resources.Requests[resourceName]; ok {
+			requestQuantity.Add(rQuantity)
```
Should rQuantity.MilliValue() be used if resourceName is v1.ResourceCPU ?
